pkg/agent/controller/egress/egress_controller.go

```diff
@@ -505,6 +531,53 @@ func (c *EgressController) realizeEgressIP(egressName, egressIP string) (uint32,
 	return ipState.mark, nil
 }
 
+func bandwidthToRateLimit(bandwidth *crdv1b1.Bandwidth) *rateLimit {
+	if bandwidth == nil {
+		return nil
+	}
+	rate, _ := resource.ParseQuantity(bandwidth.Rate)
+	burst, _ := resource.ParseQuantity(bandwidth.Burst)
+	return &rateLimit{
+		Rate:  uint32(rate.Value() / 1000),
+		Burst: uint32(burst.Value() / 1000),
+	}
+}
+
+func (c *EgressController) realizeEgressQoS(eState *egressState, egressName string, mark uint32, bandwidth *crdv1b1.Bandwidth, exist bool) error {
+	if !c.trafficShapingEnabled {
+		if bandwidth != nil {
+			klog.InfoS("Bandwidth in the Egress is ignored because OVS meters are not supported or "+
+				"trafficShaping is not enabled in Antrea-agent config.", "EgressName", egressName)
+		}
+		return nil
+	}
+	var desiredRateLimit *rateLimit
+	// QoS is desired only if the Egress is configured on this Node.
+	if mark != 0 {
+		desiredRateLimit = bandwidthToRateLimit(bandwidth)
+	}
+	// Nothing changes.
+	if eState.rateLimit.Equals(desiredRateLimit) {
+		return nil
+	}
+	// It's desired to have QoS on this Node, install/override it.
+	if desiredRateLimit != nil {
+		if err := c.ofClient.InstallEgressQoS(mark, desiredRateLimit.Rate, desiredRateLimit.Burst); err != nil {
+			return err
+		}
+		eState.rateLimit = desiredRateLimit
+		return nil
+	}
+	// It's undesired to have QoS on this Node, uninstall it.
+	if eState.rateLimit != nil {
+		if err := c.ofClient.UninstallEgressQoS(eState.mark); err != nil {
```

Comment:
I just realized that eState.mark could be reset to 0 when an EgressIP is rescheduled to another Node (due to cluster node change or node selector change), then this will be wrong. I feel we need to cache the installed mark in `rateLimit`.

Reply:
Changed.